### PR TITLE
Issue #1761 Maintain a published docs against master branch

### LINKED ISSUE #1761: Maintain a published docs against master branch
I think it would make sense to have published copy of minishift doc like http://artifacts.ci.centos.org/minishift/minishift/docs/ondemand/budhrg-docs/openshift-origin/latest/minishift/index.html under URL
```
http://artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/index.html 
```

This way we will get idea before any minishift release how documentation will look like.

#### This will happen as part of master build (Job triggered when PR got merged).

### PR DESCRIPTION
**Pull request by budhrg**
Fix #1761

Tested with job https://ci.centos.org/job/minishift-pr/2841/console after tweaking `centos_ci.sh` script go via `origin/master` branch condition.

### CODE REVIEW

**Comment by budhrg** [comment]
retest this please

**Comment by budhrg** [comment]
retest this please

**Comment by budhrg** [comment]
CentOS CI is failing due to #1773

**Comment by budhrg** [comment]
retest this please

File: docs/source/contributing/ci.adoc [hunk 1]
```diff
@@ -39,6 +39,8 @@ Thus it is used to run integration tests in addition to unit tests.
 The artifacts of a successful master build can be found at link:http://artifacts.ci.centos.org/minishift/minishift/master/[artifacts.ci.centos.org/minishift/minishift/master/<BUILD_ID>].
 The artifacts of a successful pull request build can be found at link:http://artifacts.ci.centos.org/minishift/minishift/pr/[artifacts.ci.centos.org/minishift/minishift/pr/<PR_ID>].
 
+As part of master build, a {project} documentation integrated with OpenShift documentation is hosted at link:http://artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/index.html[artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/] which reflects the documentation from the _master_ branch.
```
Comment (robin-owen) [inline docs/source/contributing/ci.adoc]: "As part of a master build, {project} documentation integrated..."

I'm also not particularly fond of having "which reflects documentation from the _master_ branch" at the end of the sentence here. It's ambiguous -- are you talking about Minishift documentation from our master branch, OpenShift documentation from their master branch, or both?

We can slightly rewrite this if I know which of those is the most technically accurate.

Comment (budhrg) [reply docs/source/contributing/ci.adoc]: @robin-owen Slightly changed the documentation. See below

screenshot from 2017-12-04 11-59-11

**Comment by budhrg** [comment]
@robin-owen @hferentschik Tested as 

### Minishift-docs job hosting documentation verification

#### Budhram Repo / docs branch (behind upstream master):
Job - https://ci.centos.org/job/minishift-pr/2894/consoleFull
Hosted at - http://artifacts.ci.centos.org/minishift/minishift/docs/ondemand/budhrg-docs/openshift-origin/latest/minishift/index.html

### Master build hosting documentation verification
Job - https://ci.centos.org/job/minishift-pr/2896/consoleFull
Hosted at - http://artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/index.html

File: centos_ci.sh [hunk 2]
```diff
@@ -293,26 +293,32 @@ function perform_release() {
...
-  RSYNC_PASSWORD=$RSYNC_PASSWORD rsync -av --delete --relative minishift/docs/ondemand/$REPO_OWNER-$BRANCH/ minishift@artifacts.ci.centos.org::minishift/
-  echo "Find docs tar here http://artifacts.ci.centos.org/minishift/minishift/docs/ondemand/$REPO_OWNER-$BRANCH/"
-  echo "Minishift documentation is hosted at http://artifacts.ci.centos.org/minishift/minishift/docs/ondemand/$REPO_OWNER-$BRANCH/openshift-origin/latest/minishift/index.html"
+  # Publish documentation for master build
+  if [[ "$JOB_NAME" = "minishift" ]]; then
+    mkdir -p minishift/docs/master
+    cp -r openshift-docs/_preview/openshift-origin minishift/docs/master/ # Copy the openshift-origin
+    # http://stackoverflow.com/a/22908437/1120530; Using --relative as --rsync-path not working
+    RSYNC_PASSWORD=$RSYNC_PASSWORD rsync -av --delete --relative minishift/docs/master/ minishift@artifacts.ci.centos.org::minishift/
+    echo "Minishift documentation is hosted at http://artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/index.html"
```
Comment (LalatenduMohanty) [inline centos_ci.sh]: I would suggest `Documentation from Minishift master branch compiled and hosted at LOCATION` . Mention of Openshift Origin does not add any value IMO

Comment (budhrg) [reply centos_ci.sh]: @robin-owen , Let me know your opinion on above comment. Does it make sense to remove OpenShift reference here at all to avoid any ambiguity or does the updated text sounds fine to you.

Comment (robin-owen) [reply centos_ci.sh]: @LalatenduMohanty @budhrg Personally, I see value in mentioning that it's integrated with the documentation for OpenShift. In this case, it's a matter of expectation, so knowing what I'm about to view is valuable. The updated text avoids the ambiguity in my opinion.

File: docs/source/contributing/ci.adoc [hunk 3]
```diff
@@ -39,6 +39,8 @@ Thus it is used to run integration tests in addition to unit tests.
 The artifacts of a successful master build can be found at link:http://artifacts.ci.centos.org/minishift/minishift/master/[artifacts.ci.centos.org/minishift/minishift/master/<BUILD_ID>].
 The artifacts of a successful pull request build can be found at link:http://artifacts.ci.centos.org/minishift/minishift/pr/[artifacts.ci.centos.org/minishift/minishift/pr/<PR_ID>].
 
+As part of master build, a {project} documentation which is compiled from the _master_ branch and integrated with OpenShift documentation is hosted at link:http://artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/index.html[artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/].
```
Comment (robin-owen) [inline docs/source/contributing/ci.adoc]: I'd slightly rewrite this opening to the following: "As part of a master build, {project} documentation"

File: centos_ci.sh [hunk 4]
```diff
@@ -322,6 +328,12 @@ function build_and_test() {
   # Run integration test with 'kvm' driver
   MINISHIFT_VM_DRIVER=kvm make integration_pr
   echo "CICO: Tests ran successfully"
+
+  # Publish documentation for master build only
+  if [[ "$GIT_BRANCH" = "origin/master" ]]; then
```
Comment (hferentschik) [inline centos_ci.sh]: What does this actually mean here? Are we now building and hosting the docs as part of each pull request? Or is this just for the push to master. If the latter, why do we handle the master push the same way as the pull request build? Really, the distinction between master and pull request build should be made at the end of the script where we switch on the job name. In the last else branch we should distinguish between master and pull request and call functions accordingly. 
In this case we don't have to have this sneaky not quite clear if statement. 

More importantly, if we have a dedicated method for the master build we can also let the full test suite run, instead of just the pull request integration test. We are giving an opportunity of our hands to get feedback early. 

Bottom line:
* The job switch statement should already determine master vs pull request build and call dedicated functions
* The function of the master build should include the docs build and publish
* On the master build we could also run all integration tests (does the nightly pass?)

File: docs/source/contributing/ci.adoc [hunk 5]
```diff
@@ -39,6 +39,8 @@ Thus it is used to run integration tests in addition to unit tests.
 The artifacts of a successful master build can be found at link:http://artifacts.ci.centos.org/minishift/minishift/master/[artifacts.ci.centos.org/minishift/minishift/master/<BUILD_ID>].
 The artifacts of a successful pull request build can be found at link:http://artifacts.ci.centos.org/minishift/minishift/pr/[artifacts.ci.centos.org/minishift/minishift/pr/<PR_ID>].
 
+As part of master build, {project} documentation which is compiled from the _master_ branch and integrated with OpenShift documentation is hosted at link:http://artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/index.html[artifacts.ci.centos.org/minishift/minishift/docs/master/openshift-origin/latest/minishift/].
```
Comment (hferentschik) [inline docs/source/contributing/ci.adoc]: 👍

File: centos_ci.sh (hunk 4 above)
Comment (budhrg) [reply centos_ci.sh]: > * The job switch statement should already determine master vs pull request build and call dedicated functions
> * The function of the master build should include the docs build and publish
> * On the master build we could also run all integration tests (does the nightly pass?)

@hferentschik  All addressed in recent refactoring.

Yes, nightly build is passing. See the last passed nightly build https://ci.centos.org/job/minishift-nightly/19/consoleFull

**Comment by budhrg** [comment]
Now, CI job is taking ~56 mins https://ci.centos.org/job/minishift-pr/2938/ after merging of https://github.com/minishift/minishift/pull/1685 (expected).

File: centos_ci.sh [hunk 6]
```diff
@@ -292,10 +289,9 @@ function perform_release() {
   curl http://minibot.19cf262c.svc.dockerapp.io:9009/hubot/centosci -H "Content-Type: application/json" -d '{"payload":{"status":"success","message":'"\"$MESSAGE\""',"url":'"\"$URL\""'}}' || true
 }
 
+# Accepts two arguments: repo owner and location in the artifact server
```
Comment (hferentschik) [inline centos_ci.sh]: Could you please document in the function docs which parameters a function expects and what they mean. See also https://google.github.io/styleguide/shell.xml#Comments

It is really hard to figure out what's going on if one just sees $1, $2, etc

File: centos_ci.sh [hunk 7]
```diff
@@ -353,14 +335,26 @@ else
...
+    REPO_OWNER=$(echo $REPO | cut -d"/" -f4)
+    perform_docs_publish "$REPO_OWNER" "minishift/docs/ondemand/$REPO_OWNER-$BRANCH";
   elif [[ "$JOB_NAME" = "minishift-release" ]]; then
-    perform_release $RSYNC_PASSWORD;
+    perform_release "$RSYNC_PASSWORD";
   elif [[ "$JOB_NAME" = "minishift-nightly" ]]; then
     perform_nightly;
-  else
-    build_and_test $RSYNC_PASSWORD;
+  elif [[ "$JOB_NAME" = "minishift" ]]; then
```
Comment (hferentschik) [inline centos_ci.sh]: jobname == minishift means it is the master build? Maybe add a comment. Otherwise much better

File: centos_ci.sh (hunk 6 above)
Comment (budhrg) [reply centos_ci.sh]: Thanks. Make sense :+1:

File: centos_ci.sh [hunk 8]
```diff
@@ -353,14 +335,26 @@ else
...
+    make prerelease synopsis_docs link_check_docs
```
Comment (hferentschik) [inline centos_ci.sh]: why don't you put these into functions like perform_master_build and perform_pr_build. That would align better with the rest of the flow.

File: centos_ci.sh (hunk 7 above)
Comment (budhrg) [reply centos_ci.sh]: Yes, master build., will add comment

File: docs/source/contributing/ci.adoc (hunk 5 above)
Comment (hferentschik) [reply docs/source/contributing/ci.adoc]: while on it, add the comment about Circle CI doing the commit message validation. I think there is a dedicated issue, so you could add it as a separate commit. While we are on it so to say

Comment (robin-owen) [reply docs/source/contributing/ci.adoc]: @hferentschik This is being handled in #1800. :)

File: centos_ci.sh [hunk 9]
```diff
@@ -255,10 +261,39 @@ function setup_build_environment() {
   runuser -l minishift_ci -c "/bin/bash centos_ci.sh"
 }
 
-function perform_release() {
-  setup_kvm_docker_machine_driver; # Required for integeration tests
-  cd gopath/src/github.com/minishift/minishift
+#############################################################################
+# Perform the publishing of Minishift documentation after integrating it with
```
Comment (budhrg) [inline centos_ci.sh]: @minishift/minishift-dev Does this comment make sense? 

Check the argument section. Is it conveying that 1 argument is Repo owner, 2nd is Branch name etc

Comment (praveenkumar) [reply centos_ci.sh]: @budhrg  How about *Publish Minishift documentation after integrating it with Openshift documentation*?

Comment (gbraad) [reply centos_ci.sh]: Not sure what you mean... but I miss checking for empty values. If any of the arguments is missing, it should fail

**Comment by budhrg** [comment]
Tested all the if-elif-else flow except release job since it separate entity and PR changes has no effect on it.

Minishift-docs flow - https://ci.centos.org/job/minishift-pr/2953/consoleFull
Minishift master job flow - https://ci.centos.org/job/minishift-pr/2949/console
Minishift Pr job flow - https://ci.centos.org/job/minishift-pr/2952/consoleFull

File: centos_ci.sh [hunk 10]
```diff
@@ -255,10 +261,39 @@ function setup_build_environment() {
   runuser -l minishift_ci -c "/bin/bash centos_ci.sh"
 }
 
-function perform_release() {
-  setup_kvm_docker_machine_driver; # Required for integeration tests
-  cd gopath/src/github.com/minishift/minishift
+#######################################################################
```
Comment (hferentschik) [inline centos_ci.sh]: :-) So much better thanks.

File: centos_ci.sh [hunk 11]
```diff
@@ -255,10 +261,39 @@ function setup_build_environment() {
...
+# Publish Minishift documentation after integrating it with
+# OpenShift documentation
+# Arguments:
+#   $1 - Repo Owner
+#   $2 - Branch name
+#   $3 - Location in the artifact server where documentation is hosted
```
Comment (hferentschik) [inline centos_ci.sh]: "Location on"

File: centos_ci.sh [hunk 12]
```diff
@@ -292,50 +327,22 @@ function perform_release() {
...
   MINISHIFT_VM_DRIVER=kvm make integration_pr
-  echo "CICO: Tests ran successfully"
-  artifacts_upload_on_pr_and_master_trigger $1;
+  perform_artifacts_upload $1;
 }
 
-function prepare_for_proxy() {
-  export INTEGRATION_PROXY_CUSTOM_PORT=8181 #needs to be an unused port
-  firewall-cmd --zone=public --add-port=$INTEGRATION_PROXY_CUSTOM_PORT/tcp;
+function perform_master() {
```
Comment (hferentschik) [inline centos_ci.sh]: 👍 Nice. Also nice imo that on master pushes we run all tests. This should give us early feedback. Do we have notifications via email configured for this job? If not, they should be.

File: docs/source/contributing/ci.adoc (hunk 5 above)
Comment (hferentschik) [reply docs/source/contributing/ci.adoc]: fair enough

**Comment by hferentschik** [comment]
@budhrg LGTM, just a small typo. You can fix it and push yourself.

File: centos_ci.sh (hunk 10 above)
Comment (budhrg) [reply centos_ci.sh]: Ya :)
